src/rust/lib/graph-editor/src/component/node/port/output.rs

```diff
@@ -0,0 +1,692 @@
+///
+/// The struct that contains the port base shapes, and some information about the shapes that is
+/// required for the computation of segments.
+struct BaseShapeData {
+    port_area  : AnyShape,
+    hover_area : AnyShape,
+    radius     : Var<Distance<Pixels>>,
+    width      : Var<Distance<Pixels>>,
+}
+
+impl BaseShapeData {
+    fn new
+    (width:&Var<Distance<Pixels>>, height:&Var<Distance<Pixels>>, grow:&Var<f32>) -> Self {
+        let width  = width  - node::NODE_SHAPE_PADDING.px() * 2.0;
+        let height = height - node::NODE_SHAPE_PADDING.px() * 2.0;
+
+        let hover_area_width  = &width  + &SHAPE_HOVER_AREA_SIZE.px() * 2.0;
+        let hover_area_height = &height / 2.0 + &SHAPE_HOVER_AREA_SIZE.px();
+        let hover_area        = Rect((&hover_area_width,&hover_area_height));
+        let hover_area        = hover_area.translate_y(-hover_area_height/2.0);
+
+        let shrink           = 1.px() - 1.px() * grow;
+        let radius           = node::NODE_SHAPE_RADIUS.px();
+        let port_area_size   = SHAPE_MAX_WIDTH.px() * grow;
+        let port_area_width  = &width  + (&port_area_size - &shrink) * 2.0;
+        let port_area_height = &height + (&port_area_size - &shrink) * 2.0;
+        let bottom_radius    = &radius + &port_area_size;
+        let port_area        = Rect((&port_area_width,&port_area_height));
+        let port_area        = port_area.corners_radius(&bottom_radius);
+        let port_area        = port_area - BottomHalfPlane();
+        let corner_radius    = &port_area_size / 2.0;
+        let corner_offset    = &port_area_width / 2.0 - &corner_radius;
+        let corner           = Circle(&corner_radius);
+        let left_corner      = corner.translate_x(-&corner_offset);
+        let right_corner     = corner.translate_x(&corner_offset);
+        let port_area        = port_area + left_corner + right_corner;
+        let port_area        = port_area.into();
+        let hover_area       = hover_area.into();
+        BaseShapeData{port_area,hover_area,radius,width}
+    }
+}
+
+/// Trait that allows us to abstract the API of the `multi_port_area::Shape` and the
+/// `single_port_area::Shape`. This is needed to avoid code duplication for functionality that can
+/// work with either shape.
+#[allow(missing_docs)]
+trait PortShape {
+    fn set_grow(&self, grow_value:f32);
+    fn set_opacity(&self, opacity:f32);
+}
+
+
+
+// ========================
+// === Multi Port Shape ===
+// ========================
+
+/// Implements the shape for a segment of the OutputPort with multiple output ports.
+pub mod multi_port_area {
+    use super::*;
+    use ensogl::display::shape::*;
+    use std::f32::consts::PI;
+
+    /// Compute the angle perpendicular to the shape border.
+    fn compute_border_perpendicular_angle
+    (full_shape_border_length:&Var<f32>, corner_segment_length:&Var<f32>, position:&Var<f32>)
+        -> Var<f32> {
```

Comment:
remove indent pls - its part of the signaturee, so indent should be the same as arg list